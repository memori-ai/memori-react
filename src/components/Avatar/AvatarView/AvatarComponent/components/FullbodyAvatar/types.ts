export enum AnimationState {
  LOADING = 'LOADING',
  EMOTION = 'EMOTION',
  IDLE = 'IDLE',
}

export interface AnimationConfig {
  fadeInDuration: number;
  fadeOutDuration: number;
  idleCount: number;
  timeScale: number;
}

export interface FullbodyAvatarProps {
  url: string;
  sex: 'MALE' | 'FEMALE';
  onLoaded?: () => void;
  currentBaseAction: {
    action: string;
    weight: number;
  };
  timeScale: number;
  onCameraZChange: (value: number) => void;
  eyeBlink?: boolean;
  avatarDepth?: number;
  avatarHeight?: number;
  stopProcessing: () => void;
  resetVisemeQueue: () => void;
  updateCurrentViseme: (
    currentTime: number
  ) => { name: string; weight: number } | null;
  smoothMorphTarget?: boolean;
  morphTargetSmoothing?: number;
  morphTargetInfluences: Record<string, number>;
  setMorphTargetDictionary: (
    morphTargetDictionary: Record<string, number>
  ) => void;
  setMorphTargetInfluences: (
    morphTargetInfluences: Record<string, number>
  ) => void;
  emotionMorphTargets: Record<string, number>;
  halfBody: boolean;
}
