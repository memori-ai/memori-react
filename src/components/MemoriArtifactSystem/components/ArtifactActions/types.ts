/**
 * TypeScript interfaces for artifact copy system components
 */

import React from 'react';

export interface CopyFormat {
  id: string;
  label: string;
  description?: string;
  icon?: React.ReactNode | undefined;
  mimeType: string;
  action: 'copy' | 'pdf' | 'print' | 'download' | 'link';
  isAsync?: boolean;
}

export interface CopyMenuItemProps {
  format: CopyFormat;
  onClick: (format: CopyFormat) => void;
  loading?: boolean;
  disabled?: boolean;
}

// CopyDropdownProps removed - using headless-ui Menu directly

export interface CopyButtonWithDropdownProps {
  artifact: {
    content: string;
    mimeType: string;
    title?: string;
  };
  onCopy?: () => void;
  onDownload?: () => void;
  onPrint?: () => void;
  onOpenExternal?: () => void;
  loading?: boolean;
  className?: string;
  disabled?: boolean;
}

export interface CopyState {
  isOpen: boolean;
  loading: boolean;
  success: boolean;
  error: string | null;
  activeFormat: string | null;
}

export interface UseCopyArtifactReturn {
  copyState: CopyState;
  formats: CopyFormat[];
  handleCopy: (format: CopyFormat) => Promise<void>;
  toggleDropdown: () => void;
  closeDropdown: () => void;
  resetState: () => void;
  handleCopyClick: () => Promise<void>;
}

export interface PDFExportOptions {
  title?: string;
  author?: string;
  margin?: {
    top: string;
    right: string;
    bottom: string;
    left: string;
  };
  fontSize?: string;
  fontFamily?: string;
  lineHeight?: string;
  color?: string;
  backgroundColor?: string;
}
