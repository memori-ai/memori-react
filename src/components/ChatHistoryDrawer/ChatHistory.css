.memori-chat-history-drawer--card {
    max-height: 60vh;
    border: 1px solid var(--memori-border);
    border-radius: 0.75rem;
    overflow-y: auto;
    transition: all 0.2s ease-in-out;
}

.memori-chat-history-drawer--card--hoverable:hover {
    border-color: var(--memori-primary-light);
    box-shadow: 0 4px 6px -1px rgba(0, 0, 0, 0.1), 0 2px 4px -1px rgba(0, 0, 0, 0.06);
}

.memori-chat-history-drawer--content {
    display: flex;
    flex-direction: column;
    gap: 1rem;
}
.memori-chat-history-drawer--toolbar--min-messages-filter{
    display: flex;
    flex-direction: row;
    align-items: center;
    gap: 0.5rem;
}

.memori-chat-history-drawer--toolbar--min-messages-filter--select{
    width: 100%;
    min-width: 240px;
}

.memori-chat-history-drawer--toolbar--min-messages-filter--input-number{
    position: relative;
    width: 100%;
    max-width: 60px;
    min-height: 32px;
    /* padding-top: 0.5rem;
    padding-right: 2.5rem;
    padding-bottom: 0.5rem;
    padding-left: 0.75rem; */
    border: 1px solid #e5e7eb;
    border-radius: 0.5rem;
    background: transparent;
    box-shadow: rgba(0, 0, 0, 0) 0px 0px 0px 0px, rgba(0, 0, 0, 0) 0px 0px 0px 0px, rgba(0, 0, 0, 0.1) 0px 4px 6px -1px,
        rgba(0, 0, 0, 0.1) 0px 2px 4px -2px;
    color: var(--memori-text-color, #333);
    cursor: pointer;
    font-family: var(--memori-font-family);
    font-size: 16px;
    text-align: left;
}

.memori-chat-history-drawer--card--header {
    position: sticky;
    z-index: 100;
    top: 0px;
    display: flex;
    flex-direction: row;
    align-items: center;
    padding: 0px;
    border-radius: 0.75rem 0.75rem 0 0;
    background-color: #ffffff;
    gap: 1rem;
}

.memori-chat-history-drawer--card--header--content {
    display: flex;
    width: 100%;
    flex-direction: row;
    align-items: center;
    gap: 1rem;
}

.memori-chat-history-drawer--card--header--info {
    display: flex;
    width: 100%;
    flex-direction: row;
    align-items: center;
    justify-content: space-between;
    gap: 0.5rem;
}

.memori-chat-history-drawer--card--header--icon-wrapper {
    display: flex;
    width: 2rem;
    height: 2rem;
    align-items: center;
    justify-content: center;
    border-radius: 0.5rem;
    background-color: var(--memori-primary-light);
    transition: all 0.2s ease-in-out;
}

.memori-chat-history-drawer--card--hoverable:hover .memori-chat-history-drawer--card--header--icon-wrapper {
    transform: scale(1.05);
}

.memori-chat-history-drawer--card--header--icon {
    color: var(--memori-primary);
}

.memori-chat-history-drawer--card--header--title-wrapper {
    display: flex;
    align-items: center;
    gap: 0.5rem;
}

.memori-chat-history-drawer--card--header--title {
    margin: 0;
    color: var(--memori-color-text);
    font-weight: 600;
}

.memori-chat-history-drawer--card--header--badge {
    padding: 0.25rem 0.5rem;
    border-radius: 9999px;
    background-color: var(--memori-primary-light);
    color: var(--memori-primary);
    font-size: 0.75rem;
    font-weight: 500;
    letter-spacing: 0.025em;
}

.memori-chat-history-drawer--card--header--meta {
    display: flex;
    align-items: center;
    margin-top: 0.25rem;
    color: var(--memori-color-text-secondary);
    font-size: 0.875rem;
    gap: 1rem;
}

.memori-chat-history-drawer--card--header--meta--messages,
.memori-chat-history-drawer--card--header--meta--time {
    display: flex;
    align-items: center;
    gap: 0.25rem;
}

.memori-chat-history-drawer--card--content {
    position: relative;
    display: flex;
    height: 100%;
    flex-direction: column;
}

.memori-chat-history-drawer--card--content--header {
    position: absolute;
    z-index: 10;
    right: 12px;
}

.memori-chat-history-drawer--card--content--export-button {
    display: flex;
    align-items: center;
    padding: 8px;
    border: none;
    background: transparent;
    color: var(--memori-primary);
    cursor: pointer;
    gap: 8px;
    transition: color 0.2s ease;
}

.memori-chat-history-drawer--card--content--export-button:hover {
    color: var(--memori-primary);
}

.memori-chat-history-drawer--card--content--export-button:disabled {
    background-color: transparent;
}

.memori-chat-history-drawer--card--content--export-button--content {
    display: flex;
    align-items: center;
    gap: 8px;
}

.memori-chat-history-drawer--card--content--export-button--icon {
    width: 18px;
    height: 18px;
    fill: var(--memori-primary);
}

.memori-chat-history-drawer--toolbar--actions > div > ul {
    z-index: 10;
    right: 0.25rem;
    margin-left: auto;
}

.memori-chat-history-drawer--card--content--messages {
    flex: 1;
    padding: 16px;
    overflow-y: auto;
}

.memori-chat-history-drawer--card--content--messages .memori-chat--cover {
    display: none;
}
.memori-chat-history-drawer--card--content--messages .memori-chat--content {
    padding: 0 !important;
    mask-image: none !important;
}

.memori-chat-history-drawer--card--content--messages--user {
    display: flex;
    flex-direction: row;
    justify-content: flex-start;
    gap: 1rem;
}

.memori-chat-history-drawer--card--content--messages--assistant {
    display: flex;
    flex-direction: row;
    justify-content: flex-end;
    gap: 1rem;
}

.memori-chat-history-drawer--card--content--actions {
    position: sticky;
    z-index: 10;
    right: 0;
    bottom: 0;
    left: 0;
    display: flex;
    justify-content: center;
    padding: 16px;
    background: linear-gradient(to top, var(--memori-background-color, #fff) 80%, transparent);
}

.memori-chat-history-drawer--card--content--resume-button {
    display: flex;
    min-width: 200px;
    align-items: center;
    justify-content: center;
    padding: 12px 24px;
    border-radius: 24px;
    gap: 8px;
    transition: transform 0.2s ease;
}

.memori-chat-history-drawer--card--content--resume-button:hover {
    transform: translateY(-2px);
}

.memori-chat-history-drawer--card--content--resume-button--content {
    display: flex;
    align-items: center;
    gap: 8px;
}

.memori-chat-history-drawer--card--content--resume-button--icon {
    width: 20px;
    height: 20px;
}

/* Loading state */
.memori-chat-history-drawer--loading {
    display: flex;
    min-height: 80px;
    flex-direction: column;
    align-items: center;
    justify-content: center;
    padding: 2rem;
    overflow-y: hidden;
}

.memori-chat-history-drawer--loading > div {
    overflow-y: hidden !important;
}

.memori-chat-history-drawer--loading--spinner {
    min-height: 80px !important;
}

.memori-chat-history-drawer--loading--text {
    margin-top: 1rem;
    color: var(--memori-color-text-secondary);
}

.memori-chat-history-drawer--card--header--meta--icon {
    width: 1.25rem;
    height: 1.25rem;
    color: var(--memori-color-text-secondary);
}
/* Empty state */
.memori-chat-history-drawer--empty {
    padding: 3rem 0;
    text-align: center;
}

.memori-chat-history-drawer--empty--icon {
    width: 3rem;
    height: 3rem;
    margin: 0 auto;
    color: var(--memori-color-text-secondary);
}

.memori-chat-history-drawer--empty--text {
    margin-top: 1rem;
    color: var(--memori-color-text-secondary);
}

/* Error state */
.memori-chat-history-drawer--error {
    padding: 1rem;
    border-radius: 0.5rem;
    background-color: var(--memori-color-error-light);
    color: var(--memori-color-error);
}

/* No results state */
.memori-chat-history-drawer--no-results {
    padding: 3rem 0;
    color: var(--memori-color-text-secondary);
    text-align: center;
}

/* Toolbar */
.memori-chat-history-drawer--toolbar {
    display: flex;
    margin-bottom: 1.5rem;
    margin-left: auto;
    gap: 1rem;
}

.memori-chat-history-drawer--toolbar--search {
    position: relative;
    flex: 1;
}

.memori-chat-history-drawer--toolbar--search--input {
    width: 100%;
    padding: 0.5rem 1rem 0.5rem 2.5rem;
    border: 1px solid var(--memori-border);
    border-radius: 0.5rem;
}

.memori-chat-history-drawer--toolbar--search--input:focus {
    border-color: var(--memori-primary);
    box-shadow: 0 0 0 2px var(--memori-primary-light);
    outline: none;
}

.memori-chat-history-drawer--toolbar--search--icon {
    position: absolute;
    top: 50%;
    left: 0.75rem;
    color: var(--memori-color-text-secondary);
    transform: translateY(-50%);
}

.memori-chat-history-drawer--toolbar--search--icon svg {
    width: 1.25rem;
    height: 1.25rem;
}

.memori-chat-history-drawer--toolbar--sort-button {
    background-color: #fff;
    color: var(--memori-primary);
    white-space: nowrap;
}

.memori-chat-history-drawer--toolbar--sort-button--icon {
    width: 1.25rem;
    height: 1.25rem;
    margin-right: 0.5rem;
    font-size: 1.25rem;
    transition: transform 0.2s ease-in-out;
}

.memori-chat-history-drawer--toolbar--sort-button--icon.rotate-180 {
    transform: scaleY(-1);
}

/* Pagination */
.memori-chat-history-drawer--pagination {
    display: flex;
    align-items: center;
    justify-content: space-between;
    margin-top: 1.5rem;
}

.memori-chat-history-drawer--pagination--button {
    width: 6rem;
}

.memori-chat-history-drawer--pagination--info {
    color: var(--memori-primary);
    font-size: 0.875rem;
}

/* Card selected state */
.memori-chat-history-drawer--card--selected {
    border-color: var(--memori-primary);
    box-shadow: 0 0 0 2px var(--memori-primary-light);
}

/* Card disabled state */
.memori-chat-history-drawer--card--disabled {
    cursor: not-allowed;
    opacity: 0.5;
}

.memori-chat-history-drawer--download-button-wrapper{
    display: flex;
    align-items: center;
    gap: 8px;
}

.memori-chat-history-drawer--download-button-wrapper--text{
    color: var(--memori-primary);
    font-size: 0.875rem;
    font-weight: 500;
}