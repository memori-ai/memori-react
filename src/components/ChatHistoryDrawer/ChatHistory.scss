.memori-chat-history-drawer {
  // ... existing code ...

  &--toolbar {
    // ... existing code ...

    &--actions {
      display: flex;
      gap: 8px;
      align-items: center;
    }

    &--sort-button {
      background: transparent;
      border: 1px solid var(--memori-primary-color);
      color: var(--memori-primary-color);

      &:hover {
        background: var(--memori-primary-color);
        color: white;
      }
    }

    &--export-button {
      background: transparent;
      border: 1px solid var(--memori-primary-color);
      color: var(--memori-primary-color);

      &:hover {
        background: var(--memori-primary-color);
        color: white;
      }

      &:disabled {
        opacity: 0.5;
        cursor: not-allowed;
      }
    }
  }
  // ... existing code ...
} 