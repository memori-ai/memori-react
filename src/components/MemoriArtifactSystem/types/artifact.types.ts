export interface ArtifactData {
  id: string;
  content: string;
  mimeType: string;
  title: string;
  timestamp: Date;
  size: number;
}

export interface ArtifactSystemState {
  currentArtifact: ArtifactData | null;
  isDrawerOpen: boolean;
  isFullscreen: boolean;
}

export type ArtifactTab = 'code' | 'preview';