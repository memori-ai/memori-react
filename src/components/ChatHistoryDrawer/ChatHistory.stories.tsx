import React from 'react';
import { Meta, StoryObj } from '@storybook/react';
import ChatHistory from './ChatHistory';
import I18nWrapper from '../../I18nWrapper';
import { ChatLog, Memori } from '@memori.ai/memori-api-client/dist/types';

// Mock API client
const mockApiClient = {
  chatLogs: {
    getUserChatLogsByToken: (loginToken: string, memoriID: string, dateFrom: string, dateTo: string) => Promise.resolve({
      chatLogs: mockChatLogs
    }),
    getUserChatLogsByTokenPaged: (requestBody: any) => Promise.resolve({
      chatLogs: mockChatLogs,
      totalItems: mockChatLogs.length
    })
  }
};

// Sample data
const mockChatLogs: ChatLog[] = [
  {
    chatLogID: 'chat123456',
    memoriID: 'mem123',
    sessionID: 'session123',
    timestamp: '2023-01-01T10:00:00Z',
    lines: [
      {
        text: 'Hello, how can I help?',
        inbound: true,
        timestamp: '2023-01-01T10:00:00Z',
        contextVars: {},
        media: [],
        memoryID: 'mem123',
      },
      {
        text: 'I need information about your services',
        inbound: false,
        timestamp: '2023-01-01T10:01:00Z',
        contextVars: {},
        media: [],
        memoryID: 'mem123',
      },
      {
        text: 'We offer a range of services including consultation, implementation, and support.',
        inbound: true,
        timestamp: '2023-01-01T10:02:00Z',
        contextVars: {},
        media: [],
        memoryID: 'mem123',
      }
    ],
    boardOfExperts: false
  },
  {
    chatLogID: 'chat456789',
    memoriID: 'mem123',
    sessionID: 'session123',
    timestamp: '2023-01-02T14:00:00Z',
    lines: [
      {
        text: 'What are your business hours?',
        inbound: false,
        timestamp: '2023-01-02T14:00:00Z',
        contextVars: {},
        media: [],
        memoryID: 'mem123',
      },
      {
        text: 'We are open from 9 AM to 5 PM weekdays.',
        inbound: true,
        timestamp: '2023-01-02T14:01:00Z',
        contextVars: {},
        media: [],
        memoryID: 'mem123',
      }
    ],
    boardOfExperts: true
  },
  {
    chatLogID: 'chat789012',
    memoriID: 'mem123',
    sessionID: 'session123',
    timestamp: '2023-01-03T11:00:00Z',
    lines: [
      {
        text: 'Do you offer remote services?',
        inbound: false,
        timestamp: '2023-01-03T11:00:00Z',
        contextVars: {},
        media: [],
        memoryID: 'mem123',
      },
      {
        text: 'Yes, we offer both in-person and remote services to accommodate your needs.',
        inbound: true,
        timestamp: '2023-01-03T11:01:00Z',
        contextVars: {},
        media: [],
        memoryID: 'mem123',
      }
    ],
    boardOfExperts: false
  },
  {
    chatLogID: 'chat345678',
    memoriID: 'mem123',
    sessionID: 'session1234',
    timestamp: '2023-01-04T09:30:00Z',
    lines: [
      {
        text: 'What is your pricing structure?',
        inbound: false,
        timestamp: '2023-01-04T09:30:00Z',
        contextVars: {},
        media: [],
        memoryID: 'mem123',
      },
      {
        text: 'Our pricing is based on project scope. We offer packages starting at $499 for basic services, with custom quotes for larger projects.',
        inbound: true,
        timestamp: '2023-01-04T09:31:00Z',
        contextVars: {},
        media: [],
        memoryID: 'mem123',
      },
      {
        text: 'Do you offer any discounts?',
        inbound: false,
        timestamp: '2023-01-04T09:32:00Z',
        contextVars: {},
        media: [],
        memoryID: 'mem123',
      },
      {
        text: 'Yes, we offer discounts for non-profits and educational institutions. We also have loyalty programs for returning customers.',
        inbound: true,
        timestamp: '2023-01-04T09:33:00Z',
        contextVars: {},
        media: [],
        memoryID: 'mem123',
      }
    ],
    boardOfExperts: false
  }
];

const mockMemori: Memori = {
  memoriID: 'mem123',
  name: 'Test Memori',
  avatar: { url: 'https://example.com/avatar.jpg' },
  description: 'A test memori',
  isPublic: true,
  integrationEnabled: false,
  enableBoardOfExperts: true,
  memoriConfigurationID: 'mem123',
  privacyType: 'PUBLIC',
  voiceType: 'FEMALE'
} as Memori;

const mockParams = {
    open: true,
    sessionId: 'session12345',
    memori: mockMemori,
    apiClient: mockApiClient as any,
    onClose: () => console.log('Close button clicked'),
    history: mockChatLogs.map(chatLog => ({
      text: chatLog.lines[0].text,
      fromUser: chatLog.lines[0].inbound,
      timestamp: chatLog.lines[0].timestamp,
    })),
    resumeSession: (chatLog: ChatLog) => {
      console.log('Resume session called with:', chatLog);
    },
    loginToken: 'mock-login-token',
    language: 'EN',
    userLang: 'EN',
}

// Create a meta object for the component
const meta: Meta<typeof ChatHistory> = {
  title: 'Widget/Chat History Drawer',
  component: ChatHistory,
  decorators: [
    Story => (
      <I18nWrapper>
        <Story />
      </I18nWrapper>
    ),
  ],
  parameters: {
    layout: 'fullscreen',
  },
  // Define common args that apply to all stories
  args: mockParams,
  // Add argTypes to configure controls in Storybook
  argTypes: {
    open: { control: 'boolean', description: 'Whether the drawer is open' },
    onClose: { action: 'closed' },
    resumeSession: { action: 'resumed session' },
  },
};

export default meta;

// Define types for our stories
type Story = StoryObj<typeof ChatHistory>;

// Default story
export const Default: Story = {
  args: {}
};

// With preselected chat
export const WithSelectedChat: Story = {
  args: {},
  play: async ({ canvasElement }) => {
    // We need to use the play function to simulate clicking on a chat
    const canvas = canvasElement.ownerDocument.body;
    setTimeout(() => {
      const chatCard = canvas.querySelector('.memori-chat-history-drawer--card');
      if (chatCard) {
        (chatCard as HTMLElement).click();
      }
    }, 300);
  }
};

// Empty state
export const EmptyState: Story = {
  args: {
    apiClient: {
      chatLogs: {
        getUserChatLogsByToken: (loginToken: string, memoriID: string, dateFrom: string, dateTo: string) => Promise.resolve({ chatLogs: [] }),
        getUserChatLogsByTokenPaged: (requestBody: any) => Promise.resolve({ chatLogs: [], totalItems: 0 })
      }
    } as any
  }
};

// Loading state
export const LoadingState: Story = {
  args: {
    apiClient: {
      chatLogs: {
        getUserChatLogsByToken: (loginToken: string, memoriID: string, dateFrom: string, dateTo: string) => new Promise(resolve => {
          // Simulate a slow network
          setTimeout(() => {
            resolve({ chatLogs: mockChatLogs });
          }, 2000);
        }),
        getUserChatLogsByTokenPaged: (requestBody: any) => new Promise(resolve => {
          // Simulate a slow network
          setTimeout(() => {
            resolve({ chatLogs: mockChatLogs, totalItems: mockChatLogs.length });
          }, 2000);
        })
      }
    } as any
  }
};

// Search state
export const WithSearch: Story = {
  args: {},
  play: async ({ canvasElement }) => {
    const canvas = canvasElement.ownerDocument.body;
    setTimeout(() => {
      const searchInput = canvas.querySelector('input[type="text"]');
      if (searchInput) {
        (searchInput as HTMLInputElement).value = 'pricing';
        (searchInput as HTMLInputElement).dispatchEvent(new Event('change', { bubbles: true }));
      }
    }, 300);
  }
};

// With pagination
export const WithPagination: Story = {
  args: {
    apiClient: {
      chatLogs: {
        getUserChatLogsByToken: (loginToken: string, memoriID: string, dateFrom: string, dateTo: string) => {
          // Create 20 mock chat logs to trigger pagination
          const manyLogs = Array.from({ length: 20 }, (_, i) => ({
            chatLogID: `chat${i}`,
            lines: [
              {
                text: `Question ${i}`,
                inbound: false,
                timestamp: `2023-01-0${(i % 9) + 1}T10:00:00Z`,
                contextVars: {},
                media: [],
                memoryID: 'mem123',
                sessionID: 'session123'
              },
              {
                text: `Answer ${i}`,
                inbound: true,
                timestamp: `2023-01-0${(i % 9) + 1}T10:01:00Z`,
                contextVars: {},
                media: [],
                memoryID: 'memori.memoriID',
                sessionID: 'session123'
              }
            ],
            boardOfExperts: i % 3 === 0
          }));
          return Promise.resolve({ chatLogs: manyLogs });
        },
        getUserChatLogsByTokenPaged: (requestBody: any) => {
          // Create 20 mock chat logs to trigger pagination
          const manyLogs = Array.from({ length: 20 }, (_, i) => ({
            chatLogID: `chat${i}`,
            lines: [
              {
                text: `Question ${i}`,
                inbound: false,
                timestamp: `2023-01-0${(i % 9) + 1}T10:00:00Z`,
                contextVars: {},
                media: [],
                memoryID: 'mem123',
                sessionID: 'session123'
              },
              {
                text: `Answer ${i}`,
                inbound: true,
                timestamp: `2023-01-0${(i % 9) + 1}T10:01:00Z`,
                contextVars: {},
                media: [],
                memoryID: 'memori.memoriID',
                sessionID: 'session123'
              }
            ],
            boardOfExperts: i % 3 === 0
          }));
          return Promise.resolve({ chatLogs: manyLogs, totalItems: 20 });
        }
      }
    } as any
  }
}
  

// With translation
export const WithTranslation: Story = {
  args: {
    ...mockParams,
    language: 'EN',
    userLang: 'IT',
    isMultilanguageEnabled: true
  }
};

// With current chat disabled
export const WithCurrentChatDisabled: Story = {
  args: {
    ...mockParams,
    sessionId: 'session1234',
  }
};