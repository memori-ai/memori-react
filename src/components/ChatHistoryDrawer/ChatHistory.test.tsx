import React from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom';
import ChatHistoryDrawer from './ChatHistory';
import memoriApiClient from '@memori.ai/memori-api-client';
import { ChatLog, ChatLogLine, Memori } from '@memori.ai/memori-api-client/dist/types';

// Mock the translation hook
jest.mock('react-i18next', () => ({
  useTranslation: () => ({
    t: (key: string) => {
      const translations: { [key: string]: string } = {
        'widget.chatHistory': 'Chat History',
        'widget.chatHistoryDescription': 'Your past conversations',
        'search': 'Search in chat history...',
        'write_and_speak.resumeButton': 'Resume chat',
        'previous': 'Previous',
        'next': 'Next'
      };
      return translations[key] || key;
    }
  })
}));

// Mock API client
const mockGetChatLogsByUser = jest.fn();
const mockApiClient = {
  chatLogs: {
    getChatLogsByUser: mockGetChatLogsByUser
  }
} as unknown as ReturnType<typeof memoriApiClient>;

// Sample data
const mockChatLogs: ChatLog[] = [
  {
    chatLogID: 'chat123',
    timestamp: '2023-01-01T10:00:00Z',
    memoriID: 'mem123',
    sessionID: 'session123',
    lines: [
      {
        text: 'Hello, how can I help?',
        inbound: true,
        timestamp: '2023-01-01T10:00:00Z',
        contextVars: {},
        media: []
      },
      {
        text: 'I need information about your services',
        inbound: false,
        timestamp: '2023-01-01T10:01:00Z',
        contextVars: {},
        media: []
      }
    ],
    boardOfExperts: false
  },
  {
    chatLogID: 'chat456',
    timestamp: '2023-01-02T14:00:00Z',
    memoriID: 'mem123',
    sessionID: 'session123',
    lines: [
      {
        text: 'What are your business hours?',
        inbound: false,
        timestamp: '2023-01-02T14:00:00Z',
        contextVars: {},
        media: []
      },
      {
        text: 'We are open from 9 AM to 5 PM weekdays.',
        inbound: true,
        timestamp: '2023-01-02T14:01:00Z',
        contextVars: {},
        media: []
      }
    ],
    boardOfExperts: true
  }
];

const mockMemori: Memori = {
  memoriID: 'mem123',
  name: 'Test Memori',
  avatar: { url: 'https://example.com/avatar.jpg' },
  description: 'A test memori',
  isPublic: true,
  memoriConfigurationID: 'mem123',
  privacyType: 'PUBLIC',
  voiceType: 'FEMALE'
} as Memori;

describe('ChatHistoryDrawer', () => {
  beforeEach(() => {
    mockGetChatLogsByUser.mockResolvedValue({ chatLogs: mockChatLogs });
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  test('renders the chat history drawer with correct title', async () => {
    render(
      <ChatHistoryDrawer
        open={true}
        onClose={() => {}}
        apiClient={mockApiClient}
        sessionId="session123"
        memori={mockMemori}
        resumeSession={() => {}}
         baseUrl="https://www.aisuru.com"
        apiUrl="https://backend.memori.ai"
      />
    );

    expect(screen.getByText('Chat History')).toBeInTheDocument();
    expect(screen.getByPlaceholderText('Search in chat history...')).toBeInTheDocument();
    
    // Wait for the chat logs to be fetched and displayed
    await waitFor(() => {
      expect(mockGetChatLogsByUser).toHaveBeenCalledWith('session123');
      expect(screen.getByText('Chat-chat')).toBeInTheDocument(); // Shows shortened ID
    });
  });

  test('allows searching through chat logs', async () => {
    render(
      <ChatHistoryDrawer
        open={true}
        onClose={() => {}}
        apiClient={mockApiClient}
        sessionId="session123"
        memori={mockMemori}
        resumeSession={() => {}}
        baseUrl="https://www.aisuru.com"
        apiUrl="https://backend.memori.ai"
      />
    );

    await waitFor(() => {
      expect(screen.getByText('Chat-chat')).toBeInTheDocument();
    });

    // Search for something that should filter results
    const searchInput = screen.getByPlaceholderText('Search in chat history...');
    fireEvent.change(searchInput, { target: { value: 'business hours' } });

    // Only the chat with "business hours" should remain
    await waitFor(() => {
      expect(screen.queryByText('Chat-chat')).toBeInTheDocument();
      // We should only see the one matching the search
      expect(screen.getAllByText(/Chat-/i)).toHaveLength(1);
    });
  });

  test('displays chat details when a chat log is selected', async () => {
    render(
      <ChatHistoryDrawer
        open={true}
        onClose={() => {}}
        apiClient={mockApiClient}
        sessionId="session123"
        memori={mockMemori}
        resumeSession={() => {}}
        baseUrl="https://www.aisuru.com"
        apiUrl="https://backend.memori.ai"
      />
    );

    await waitFor(() => {
      expect(screen.getByText('Chat-chat')).toBeInTheDocument();
    });

    // Click on a chat log
    fireEvent.click(screen.getByText('Chat-chat'));

    // Check that chat details are displayed
    await waitFor(() => {
      expect(screen.getByText('Hello, how can I help?')).toBeInTheDocument();
      expect(screen.getByText('I need information about your services')).toBeInTheDocument();
      expect(screen.getByText('Resume chat')).toBeInTheDocument();
    });
  });

  test('calls resumeSession when resume button is clicked', async () => {
    const mockResumeSession = jest.fn();
    render(
      <ChatHistoryDrawer
        open={true}
        onClose={() => {}}
        apiClient={mockApiClient}
        sessionId="session123"
        memori={mockMemori}
        resumeSession={mockResumeSession}
        baseUrl="https://www.aisuru.com"
        apiUrl="https://backend.memori.ai"
      />
    );

    await waitFor(() => {
      expect(screen.getByText('Chat-chat')).toBeInTheDocument();
    });

    // Click on a chat log
    fireEvent.click(screen.getByText('Chat-chat'));

    // Wait for chat details to appear
    await waitFor(() => {
      expect(screen.getByText('Resume chat')).toBeInTheDocument();
    });

    // Click resume button
    fireEvent.click(screen.getByText('Resume chat'));

    // Check that resumeSession was called with correct args
    expect(mockResumeSession).toHaveBeenCalledWith(
      'session123',
      expect.any(Array),
      expect.any(Array)
    );
  });

  test('paginates chat logs correctly', async () => {
    // Create more mock data to test pagination
    const manyLogs = Array(15).fill(null).map((_, i) => ({
      chatLogID: `chat${i}`,
      lines: [
        {
          text: `Message ${i}`,
          inbound: i % 2 === 0,
          timestamp: `2023-01-0${i % 9 + 1}T10:00:00Z`,
          contextVars: {},
          media: []
        },
        {
          text: `Response ${i}`,
          inbound: i % 2 !== 0,
          timestamp: `2023-01-0${i % 9 + 1}T10:01:00Z`,
          contextVars: {},
          media: []
        }
      ],
      boardOfExperts: i % 3 === 0
    }));
    
    mockGetChatLogsByUser.mockResolvedValue({ chatLogs: manyLogs });

    render(
      <ChatHistoryDrawer
        open={true}
        onClose={() => {}}
        apiClient={mockApiClient}
        sessionId="session123"
        memori={mockMemori}
        resumeSession={() => {}}
        apiUrl="https://backend.memori.ai"
        baseUrl="https://www.aisuru.com"
      />
    );

    // Wait for the pagination to be displayed
    await waitFor(() => {
      expect(screen.getByText('1 / 2')).toBeInTheDocument();
      expect(screen.getByText('Next')).toBeInTheDocument();
    });

    // Click next page
    fireEvent.click(screen.getByText('Next'));

    // Check we're on page 2
    await waitFor(() => {
      expect(screen.getByText('2 / 2')).toBeInTheDocument();
    });

    // Click previous page
    fireEvent.click(screen.getByText('Previous'));

    // Check we're back on page 1
    await waitFor(() => {
      expect(screen.getByText('1 / 2')).toBeInTheDocument();
    });
  });
});