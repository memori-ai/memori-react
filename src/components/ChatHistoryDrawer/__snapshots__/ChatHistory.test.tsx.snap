// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders ChatHistoryDrawer closed unchanged 1`] = `<div />`;

exports[`renders ChatHistoryDrawer unchanged 1`] = `
<div>
  <div
    style="position: fixed; top: 1px; left: 1px; width: 1px; height: 0px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; border-width: 0px; display: none;"
  />
</div>
`;

exports[`renders ChatHistoryDrawer with chat logs unchanged 1`] = `
<div>
  <div
    style="position: fixed; top: 1px; left: 1px; width: 1px; height: 0px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; border-width: 0px; display: none;"
  />
</div>
`;

exports[`renders ChatHistoryDrawer with selected chat log unchanged 1`] = `
<div>
  <div
    style="position: fixed; top: 1px; left: 1px; width: 1px; height: 0px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; border-width: 0px; display: none;"
  />
</div>
`;
